crates/amaru-ledger/src/rules/block.rs

```diff
@@ -26,64 +26,74 @@ use crate::{
-
-    #[error("Invalid block's execution units: {0}")]
-    ExUnits(#[from] InvalidExUnits),
-
-    #[error("Invalid block header: {0}")]
-    Header(#[from] InvalidBlockHeader),
-
-    #[error(
-        "Invalid transaction (hash: {transaction_hash}, index: {transaction_index}): {violation} "
-    )]
+    Size(InvalidBlockSize),
+    ExUnits(InvalidExUnits),
+    Header(InvalidBlockHeader),
     Transaction {
         transaction_hash: Hash<32>,
         transaction_index: u32,
         violation: InvalidTransaction,
     },
-
-    // TODO: This error shouldn't exist, it's a placeholder for better error handling in less straight forward cases
-    #[error("Uncategorized error: {0}")]
     UncategorizedError(String),
 }
 
+pub enum BlockValidation {
+    Valid,
+    Invalid(InvalidBlock),
+}
+
+impl Try for BlockValidation {
+    type Output = ();
+    type Residual = BlockValidationResidual;
+
+    fn from_output((): Self::Output) -> Self {
+        BlockValidation::Valid
+    }
+
+    fn branch(self) -> ControlFlow<Self::Residual, Self::Output> {
+        match self {
+            BlockValidation::Valid => ControlFlow::Continue(()),
+            BlockValidation::Invalid(e) => ControlFlow::Break(BlockValidationResidual::Invalid(e)),
+        }
+    }
+}
+
+pub enum BlockValidationResidual {
+    Invalid(InvalidBlock),
+}
+
+impl FromResidual for BlockValidation {
+    fn from_residual(residual: BlockValidationResidual) -> Self {
+        match residual {
+            BlockValidationResidual::Invalid(e) => BlockValidation::Invalid(e),
+        }
+    }
+}
+
 #[instrument(level = Level::TRACE, skip_all)]
 pub fn execute<C: ValidationContext<FinalState = S>, S: From<C>>(
-    mut context: C,
+    context: &mut C,
     protocol_params: ProtocolParameters,
     block: MintedBlock<'_>,
-) -> Result<S, InvalidBlock> {
-    header_size::block_header_size_valid(block.header.raw_cbor(), &protocol_params)?;
-
-    body_size::block_body_size_valid(&block.header.header_body, &block)?;
-
-    // TODO: rewrite this to use iterators defined on `Redeemers` and `MaybeIndefArray`, ideally
-    let ex_units = block
-        .transaction_witness_sets
-        .iter()
-        .flat_map(|witness_set| {
-            witness_set
-                .redeemer
-                .iter()
-                .map(|redeemers| match redeemers.deref() {
-                    Redeemers::List(list) => list.iter().map(|r| r.ex_units).collect::<Vec<_>>(),
-                    Redeemers::Map(map) => map.iter().map(|(_, r)| r.ex_units).collect::<Vec<_>>(),
-                })
-        })
-        .flatten()
-        .collect::<Vec<_>>();
-
-    ex_units::block_ex_units_valid(ex_units, &protocol_params)?;
+) -> BlockValidation {
+    // Block level validations functions share the same signature.
+    // Currently apply them one by one
+    // TODO consider an abstract strategy pattern to apply them (e.g. in parallel, in a priority order, ...)
+    let block_validation_fns = vec![
```

Comment:
This is an nice side-effect of having unified signature for validation fns

Reply:
Maybe having a proper trait for that would alleviate this? I think it opens the door to nice optimizations down the line.